File: scripts/backfill_rsvp_profiles.js
#!/usr/bin/env node

/**
 * One-time migration: replaces placeholder 'Current User' RSVP attendee
 * profiles with real user data from the `users` collection.
 *
 * Usage:
 *   node scripts/backfill_rsvp_profiles.js            # dry-run (default)
 *   node scripts/backfill_rsvp_profiles.js --apply     # write changes
 *
 * Auth: uses Firebase CLI cached credentials (~/.config/configstore/firebase-tools.json)
 *       or GOOGLE_APPLICATION_CREDENTIALS env var if set.
 *
 * Affected collections:
 *   event_attendees/{eventId}/attendees/{userId}
 *   user_rsvps/{userId}/events/{eventId}
 */

const { Firestore } = require("@google-cloud/firestore");
const { OAuth2Client } = require("google-auth-library");
const fs = require("fs");
const path = require("path");

const DRY_RUN = !process.argv.includes("--apply");
const BATCH_LIMIT = 400;
const PROJECT_ID = "naijasingles-74a75";

// Firebase CLI OAuth client ID (public, used by all Firebase CLI installations)
const FIREBASE_CLIENT_ID =
  "563584335869-fgrhgmd47bqnekij5i8b5pr03ho849e6.apps.googleusercontent.com";
const FIREBASE_CLIENT_SECRET = "j9iVZfS8kkCEFUPaAeJV0sAi";

function createFirestore() {
  // If GOOGLE_APPLICATION_CREDENTIALS is set, Firestore uses it automatically
  if (process.env.GOOGLE_APPLICATION_CREDENTIALS) {
    return new Firestore({ projectId: PROJECT_ID });
  }

  // Otherwise use Firebase CLI cached tokens
  const tokenPath = path.join(
    process.env.HOME,
    ".config",
    "configstore",
    "firebase-tools.json"
  );

  if (!fs.existsSync(tokenPath)) {
    console.error("❌ No Firebase credentials found.");
    console.error("   Run 'firebase login' first, or set GOOGLE_APPLICATION_CREDENTIALS.");
    process.exit(1);
  }

  const config = JSON.parse(fs.readFileSync(tokenPath, "utf8"));
  const tokens = config.tokens;

  if (!tokens || !tokens.refresh_token) {
    console.error("❌ Firebase CLI token missing or expired. Run 'firebase login'.");
    process.exit(1);
  }

  const oauth2Client = new OAuth2Client(
    FIREBASE_CLIENT_ID,
    FIREBASE_CLIENT_SECRET
  );
  oauth2Client.setCredentials({
    refresh_token: tokens.refresh_token,
  });

  return new Firestore({
    projectId: PROJECT_ID,
    authClient: oauth2Client,
  });
}

async function main() {
  console.log("🔧 RSVP Profile Backfill Migration");
  console.log("===================================");
  console.log(
    `Mode: ${DRY_RUN ? "DRY RUN (no writes)" : "⚠️  APPLY (writing to Firestore)"}`
  );
  console.log("");

  const db = createFirestore();

  // Quick connectivity check
  console.log("🔌 Connecting to Firestore...");
  try {
    await db.collection("users").limit(1).get();
    console.log("   ✅ Connected.\n");
  } catch (e) {
    console.error("   ❌ Connection failed:", e.message);
    process.exit(1);
  }

  const userProfileCache = new Map();

  async function fetchUserProfile(userId) {
    if (userProfileCache.has(userId)) return userProfileCache.get(userId);
    try {
      const doc = await db.collection("users").doc(userId).get();
      if (!doc.exists) {
        userProfileCache.set(userId, null);
        return null;
      }
      const data = doc.data();
      const photos = Array.isArray(data.photos) ? data.photos : [];
      const profile = {
        name: (data.name || "Unknown").toString(),
        avatar: photos.length > 0 ? photos[0] : null,
        age: data.age || null,
        location: data.address ? data.address.toString() : null,
      };
      userProfileCache.set(userId, profile);
      return profile;
    } catch (e) {
      console.error(`  ❌ Error fetching user ${userId}:`, e.message);
      userProfileCache.set(userId, null);
      return null;
    }
  }

  function needsBackfill(data) {
    const profile = data.userProfile;
    if (!profile || typeof profile !== "object") return true;
    const name = (profile.name || "").toString();
    return name === "" || name === "Current User" || name === "Unknown";
  }

  // --- Phase 1: Scan event_attendees ---
  console.log("📡 Scanning event_attendees...");
  const badRecords = [];

  const eventAttendeeDocs = await db.collection("event_attendees").listDocuments();
  console.log(`   Found ${eventAttendeeDocs.length} event(s) with attendees.`);

  for (const eventDocRef of eventAttendeeDocs) {
    const eventId = eventDocRef.id;
    const attendeesSnap = await eventDocRef.collection("attendees").get();

    for (const doc of attendeesSnap.docs) {
      const data = doc.data();
      if (!needsBackfill(data)) continue;

      const userId = (data.userId || "").toString();
      if (!userId) continue;

      badRecords.push({
        userId,
        eventId,
        currentName: (data.userProfile?.name || "<null>").toString(),
        docPath: doc.ref.path,
      });
    }
  }

  console.log(`   🔍 Found ${badRecords.length} records needing backfill.\n`);

  if (badRecords.length === 0) {
    console.log("✅ No bad records found. Nothing to do.");
    process.exit(0);
  }

  // --- Phase 2: Backfill ---
  console.log("🔄 Backfilling profiles...\n");

  let updated = 0;
  let skipped = 0;
  let errors = 0;

  for (let i = 0; i < badRecords.length; i += BATCH_LIMIT / 2) {
    const chunk = badRecords.slice(i, i + BATCH_LIMIT / 2);
    const batch = db.batch();
    let batchWrites = 0;

    for (const record of chunk) {
      const profile = await fetchUserProfile(record.userId);

      if (!profile) {
        console.log(`  ⏭️  SKIP ${record.userId} (user doc not found)`);
        skipped++;
        continue;
      }

      const newName = profile.name || "Unknown";
      console.log(
        `  ${DRY_RUN ? "🔍" : "✏️"} "${record.currentName}" → "${newName}" ` +
          `(user: ${record.userId}, event: ${record.eventId})`
      );

      if (!DRY_RUN) {
        const attendeeRef = db
          .collection("event_attendees")
          .doc(record.eventId)
          .collection("attendees")
          .doc(record.userId);
        batch.set(attendeeRef, { userProfile: profile }, { merge: true });

        const userRsvpRef = db
          .collection("user_rsvps")
          .doc(record.userId)
          .collection("events")
          .doc(record.eventId);
        batch.set(userRsvpRef, { userProfile: profile }, { merge: true });

        batchWrites += 2;
      }

      updated++;
    }

    if (!DRY_RUN && batchWrites > 0) {
      try {
        await batch.commit();
        console.log(`   ✅ Batch committed (${batchWrites} writes)`);
      } catch (e) {
        console.error(`   ❌ Batch error:`, e.message);
        errors += chunk.length;
        updated -= chunk.length;
      }
    }
  }

  console.log("\n===================================");
  console.log("📊 Results:");
  console.log(`   Total bad records:  ${badRecords.length}`);
  console.log(`   ${DRY_RUN ? "Would update" : "Updated"}:      ${updated}`);
  console.log(`   Skipped (no user): ${skipped}`);
  if (errors > 0) console.log(`   Errors:             ${errors}`);
  console.log(`   Mode:               ${DRY_RUN ? "DRY RUN" : "APPLIED"}`);

  if (DRY_RUN && updated > 0) {
    console.log("\n💡 Run with --apply to write changes:");
    console.log("   node scripts/backfill_rsvp_profiles.js --apply");
  }

  console.log("");
  process.exit(0);
}

main().catch((e) => {
  console.error("💥 Fatal error:", e.message || e);
  process.exit(1);
});

File: scripts/backfill_rsvp_profiles.dart
import 'dart:io';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';

final FirebaseOptions _firebaseOptions = FirebaseOptions(
  apiKey: 'AIzaSyAwsU8j3acGo_cKOECbgsXHd3-qvvLn_Fw',
  appId: '1:888697307756:ios:95ea92b8c7288e31704e49',
  messagingSenderId: '888697307756',
  projectId: 'naijasingles-74a75',
  storageBucket: 'naijasingles-74a75.appspot.com',
  iosBundleId: 'com.naijasingles.app',
);

/// One-time migration: replaces placeholder 'Current User' RSVP attendee
/// profiles with real user data from the `users` collection.
///
/// Usage:
///   dart run scripts/backfill_rsvp_profiles.dart          # dry-run (default)
///   dart run scripts/backfill_rsvp_profiles.dart --apply   # write changes
///
/// Affected collections:
///   event_attendees/{eventId}/attendees/{userId}
///   user_rsvps/{userId}/events/{eventId}
void main(List<String> args) async {
  final dryRun = !args.contains('--apply');

  print('🔧 RSVP Profile Backfill Migration');
  print('===================================');
  print('Mode: ${dryRun ? "DRY RUN (no writes)" : "⚠️  APPLY (writing to Firestore)"}');
  print('');

  await Firebase.initializeApp(options: _firebaseOptions);
  final firestore = FirebaseFirestore.instance;

  final userProfileCache = <String, Map<String, dynamic>?>{};

  Future<Map<String, dynamic>?> fetchUserProfile(String userId) async {
    if (userProfileCache.containsKey(userId)) {
      return userProfileCache[userId];
    }
    try {
      final doc = await firestore.collection('users').doc(userId).get();
      if (!doc.exists) {
        userProfileCache[userId] = null;
        return null;
      }
      final data = doc.data()!;
      final photos = data['photos'];
      final profile = {
        'name': data['name']?.toString() ?? 'Unknown',
        'avatar': (photos is List && photos.isNotEmpty) ? photos[0] : null,
        'age': data['age'],
        'location': data['address']?.toString(),
      };
      userProfileCache[userId] = profile;
      return profile;
    } catch (e) {
      print('  ❌ Error fetching user $userId: $e');
      userProfileCache[userId] = null;
      return null;
    }
  }

  bool needsBackfill(Map<String, dynamic> data) {
    final profile = data['userProfile'];
    if (profile == null) return true;
    if (profile is! Map) return true;
    final name = profile['name']?.toString() ?? '';
    return name.isEmpty || name == 'Current User' || name == 'Unknown';
  }

  // --- Phase 1: Scan event_attendees via collectionGroup ---
  print('📡 Scanning event_attendees (collectionGroup: attendees)...');

  final badRecords = <_BadRecord>[];

  try {
    final attendeesQuery = firestore.collectionGroup('attendees');
    final snapshot = await attendeesQuery.get();

    print('   Found ${snapshot.docs.length} total attendee records.');

    for (final doc in snapshot.docs) {
      final data = doc.data();
      if (!needsBackfill(data)) continue;

      final userId = data['userId']?.toString() ?? '';
      final eventId = data['eventId']?.toString() ?? '';
      if (userId.isEmpty || eventId.isEmpty) continue;

      final currentName = data['userProfile']?['name']?.toString() ?? '<null>';
      badRecords.add(_BadRecord(
        userId: userId,
        eventId: eventId,
        currentName: currentName,
        docPath: doc.reference.path,
      ));
    }
  } catch (e) {
    print('   ⚠️  collectionGroup query failed: $e');
    print('   This likely needs a Firestore index. Falling back to manual scan...');
    await _manualScan(firestore, badRecords, needsBackfill);
  }

  print('   🔍 Found ${badRecords.length} records needing backfill.\n');

  if (badRecords.isEmpty) {
    print('✅ No bad records found. Nothing to do.');
    exit(0);
  }

  // --- Phase 2: Backfill ---
  print('🔄 Backfilling profiles...\n');

  var updated = 0;
  var skipped = 0;
  var errors = 0;

  for (final record in badRecords) {
    final profile = await fetchUserProfile(record.userId);

    if (profile == null) {
      print('  ⏭️  SKIP ${record.userId} (user doc not found)');
      skipped++;
      continue;
    }

    final newName = profile['name'] ?? 'Unknown';
    print('  ${dryRun ? "🔍" : "✏️"} ${record.currentName} → $newName '
        '(user: ${record.userId}, event: ${record.eventId})');

    if (!dryRun) {
      try {
        final batch = firestore.batch();

        final attendeeRef = firestore
            .collection('event_attendees')
            .doc(record.eventId)
            .collection('attendees')
            .doc(record.userId);
        batch.update(attendeeRef, {'userProfile': profile});

        final userRsvpRef = firestore
            .collection('user_rsvps')
            .doc(record.userId)
            .collection('events')
            .doc(record.eventId);
        batch.update(userRsvpRef, {'userProfile': profile});

        await batch.commit();
        updated++;
      } catch (e) {
        print('    ❌ Error updating: $e');
        errors++;
      }
    } else {
      updated++;
    }
  }

  print('\n===================================');
  print('📊 Results:');
  print('   Total bad records:  ${badRecords.length}');
  print('   ${dryRun ? "Would update" : "Updated"}:      $updated');
  print('   Skipped (no user): $skipped');
  if (errors > 0) print('   Errors:             $errors');
  print('   Mode:               ${dryRun ? "DRY RUN" : "APPLIED"}');

  if (dryRun && updated > 0) {
    print('\n💡 Run with --apply to write changes:');
    print('   dart run scripts/backfill_rsvp_profiles.dart --apply');
  }

  print('');
  exit(0);
}

/// Fallback: manually iterate top-level event_attendees docs if
/// collectionGroup query fails (missing index).
Future<void> _manualScan(
  FirebaseFirestore firestore,
  List<_BadRecord> badRecords,
  bool Function(Map<String, dynamic>) needsBackfill,
) async {
  print('   Scanning event_attendees top-level docs...');
  final eventAttendeeDocs =
      await firestore.collection('event_attendees').get();

  for (final eventDoc in eventAttendeeDocs.docs) {
    final eventId = eventDoc.id;
    final attendees = await firestore
        .collection('event_attendees')
        .doc(eventId)
        .collection('attendees')
        .get();

    for (final doc in attendees.docs) {
      final data = doc.data();
      if (!needsBackfill(data)) continue;

      final userId = data['userId']?.toString() ?? '';
      if (userId.isEmpty) continue;

      badRecords.add(_BadRecord(
        userId: userId,
        eventId: eventId,
        currentName: data['userProfile']?['name']?.toString() ?? '<null>',
        docPath: doc.reference.path,
      ));
    }
  }
}

class _BadRecord {
  _BadRecord({
    required this.userId,
    required this.eventId,
    required this.currentName,
    required this.docPath,
  });

  final String userId;
  final String eventId;
  final String currentName;
  final String docPath;
}
